modules/031-linstor/images/linstor-pools-importer/main.go

```diff
@@ -0,0 +1,386 @@
+		if err != nil {
+			klog.Fatalln("Failed to get LINSTOR node", err)
+		}
+		_, err = lc.Nodes.GetStoragePool(ctx, nodeName, storagePool.StoragePoolName)
+		if err != nil && err == lclient.NotFoundError {
+			err = lc.Nodes.CreateStoragePool(ctx, nodeName, storagePool)
+			if err != nil {
+				sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeWarning, "Failed", "Failed to create LINSTOR storage pool"+err.Error())
+				klog.Fatalln("Failed to create LINSTOR storage pool", err)
+			}
+			sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeNormal, "Created", "Created LINSTOR storage pool: "+nodeName+"/"+storagePool.StoragePoolName)
+		} else if err != nil {
+			klog.Fatalln("Failed to get LINSTOR storage pool", err)
+		}
+
+		// Get the maximum number of available replicas
+		opts := lclient.ListOpts{StoragePool: []string{storagePool.StoragePoolName}}
+		sp, err := lc.Nodes.GetStoragePoolView(ctx, &opts)
+		if err != nil {
+			klog.Fatalln("Failed to list LINSTOR storage pools", err)
+		}
+		replicasNum := len(sp)
+		if replicasNum > maxReplicasNum {
+			replicasNum = maxReplicasNum
+		}
+
+		// Create StorageClasses in Kubernetes
+		for r := 1; r <= replicasNum; r++ {
+			storageClass := newKubernetesStorageClass(&storagePool, r)
+			err = kc.Get(ctx, types.NamespacedName{Name: storageClass.GetName()}, &storageClass)
+			if err != nil && errors.IsNotFound(err) {
+				err = kc.Create(ctx, &storageClass)
+				if err != nil {
+					sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeWarning, "Failed", "Failed to create Kubernetes storage class"+err.Error())
+					klog.Fatalln("Failed to create Kubernetes storageClass", err)
+				}
+				sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeNormal, "Created", "Created Kubernetes storage class: "+storageClass.Name)
+			} else if err != nil {
+				klog.Fatalln("Failed to get Kubernetes storage class:", err)
+			}
+		}
+	}
+}
+
+// Sends event to Kubernetes
+func sendKubernetesEvent(ctx context.Context, kc kclient.Client, owner *v1.ObjectReference, eventType, reason, message string) {
+	event := newKubernetesEvent(owner, eventType, reason, message)
+	err := kc.Create(ctx, &event)
+	if err != nil {
+		klog.Fatalln("Failed to create event", err)
+	}
+}
+
+// Makes loop over storage pool candidates, retruns channel of changed ones
+func runCandidatesLoop(ctx context.Context, f func() ([]lclient.StoragePool, error), delay time.Duration) <-chan lclient.StoragePool {
+	ch := make(chan lclient.StoragePool)
+	go func() {
+		var oldCandidates []lclient.StoragePool
+		for {
+			select {
+			case <-ctx.Done():
+				close(ch)
+				return
+			default:
+				candidates, err := f()
+				if err != nil {
+					klog.Fatalln("failed to load candidates:", err)
+				}
+			LOOP:
+				for _, candidate := range candidates {
+					for _, oldCandidate := range oldCandidates {
+						if candidate.StoragePoolName == oldCandidate.StoragePoolName {
+							continue LOOP
+						}
+					}
+					ch <- candidate
+				}
+				oldCandidates = candidates
```

Comment:
> 1. How do we detect that a pool has dissappeared? should we delete corresponding storage class then

No, because it can happen at any time when `lvs` can't read configuration from the backing device eg. in case of storage failure and so on

> 2. How do we detect that a pool has changed its configuration, e.g. fixed a name and became valid?

We have not much to do here. The only configuration may change is the name, if name is changed then pool considered as a new pool and must be created.

> 3. What if storage pools have similar names? How should we provision storage classes for them?

 First come, first serve. The first one will be configured the second one will be ignored.

> 4. What if storage pool names collide on different set of nodes? They meet only in Kubernetes cluster, and still can collide. Or do i miss something?

Yeah, the storage pools should have the same names across all the nodes, example:

```
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
┊ StoragePool ┊ Node           ┊ Driver   ┊ PoolName          ┊ FreeCapacity ┊ TotalCapacity ┊ CanSnapshots ┊ State ┊ SharedName ┊
╞════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╡
┊ lvmthin     ┊ hf-kubevirt-01 ┊ LVM_THIN ┊ linstor_data/data ┊     1.69 TiB ┊      1.80 TiB ┊ True         ┊ Ok    ┊            ┊
┊ lvmthin     ┊ hf-kubevirt-02 ┊ LVM_THIN ┊ linstor_data/data ┊     1.69 TiB ┊      1.80 TiB ┊ True         ┊ Ok    ┊            ┊
┊ lvmthin     ┊ hf-kubevirt-03 ┊ LVM_THIN ┊ linstor_data/data ┊     1.69 TiB ┊      1.80 TiB ┊ True         ┊ Ok    ┊            ┊
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

Reply:
> 1. No, because it can happen at any time when lvs can't read configuration from the backing device eg. in case of storage failure and so on

What if the configuration is read succesfully?

 > 2. We have not much to do here. The only configuration may change is the name, if name is changed then pool considered as a new pool and must be created.

Good point.

> 3. First come, first serve. The first one will be configured the second one will be ignored.

I believe we can avoid making such a choice. It is possible to hash the configuration to distinguish separate pools with the same name. For now, I think, it is more an unnecessary complication. Let's leave it this way.

> 4. Yeah, the storage pools should have the same names across all the nodes, example:

Can we have a storage called "data" on one nodegroup "worker" and another separate storage called "data" on nodegroup "database"? Can it work this way?

Reply:
> What if the configuration is read succesfully and we detect the pool is absent? How can user delete a pool then?

For now there is no logic for deleting pool from LINSTOR at this level. This code is used only to automatically detect pools on the node and add configuration for them into LINSTOR and Kubernetes.

If pool is unexpectedly absent then LINSTOR will report an error. This will always require manual actions using CLI to diagnose and recover pool and all DRBD replicas located on it. See the [example alert](https://github.com/deckhouse/deckhouse/blob/7809bef48b83a537929a70a52a0145b102532e05/modules/031-linstor/monitoring/prometheus-rules/linstor-storage-pools.yaml#L15-L19) for this.

Another words the only reason of this program is to simplify initial configuration for the end user. Because we don't want overwhelm them by the information of LINSTOR internal resources and the configuration logic.

@distol is convinced that most of our users has way more experience of using `lvm` commands and implementing this will simplify the configuration at least for the beginning. But I admit that knowledge of using LINSTOR CLI will be needed for solving issues sooner or later. However these operations can be done by the special storage team from our side, or by more experienced users after reading [ADVANCED_USAGE](https://github.com/deckhouse/deckhouse/blob/linstor-lvm-importer/modules/031-linstor/docs/ADVANCED_USAGE.md) guide 

> Can we have a storage called "data" on one nodegroup "worker" and another separate storage called "data" on nodegroup "database"? Can it work this way?

Yeah, this is main course of action. You always need to create storagepools with the same names on different nodes to have an opportunity creating persistent volumes in multiple replicas, example you have five nodes:
```
.
├── node1
├── node2
└── node3
└── node4
└── node5
```

Then you create storage pools on them like this:
```
.
├── node1
│   └── pool1
├── node2
│   └── pool1
├── node3
│   └── pool1
├── node4
│   ├── pool1
│   └── pool2
└── node5
    ├── pool1
    └── pool2
```

This will add 5 storage pools with name `pool1` (one for each node) and 2 with name `pool2` (for node 4 and 5) in LINSTOR and generate five storageclasses in Kubernetes:

```
linstor-pool1-r1
linstor-pool1-r2
linstor-pool1-r3
linstor-pool2-r1
linstor-pool2-r2
```

Using each of them will allow user to place needed amount of replicas on these nodes.

If you didn't get this, I would need to update module documentation then.

Reply:
> For now there is no logic for deleting pool from LINSTOR at this level

Ok, so this is not a part of the task. I think it worth mentioning in comments in code, which actual use-cases we decide to support and which not. 

In general, storage pool provisioning would be expected to cover not only creation, but also a cleanup for any human-implied reason, e.g. misconfiguration (name typo), or cleaning space for another storage tasks. Meaning, pools sometimes are need to be deleted. It's fine if we don't cover them. I'd prefer to mention this explicitly somewhere.

> If you didn't get this, I would need to update module documentation then.

I see that I didn't manage to ask my question clearly enough. But I probably got the answer. So, a pool name is unique cluster-wide. And the replica maximum is hardcoded. So there will always be three replicas. And extra pools with the same name are going to be ignored. It is clear


```
.
├── node1
│   └── name1   # pool 1
├── node2
│   └── name1   # pool 1
├── node3
│   └── name1   # pool 1
├── node4
│   └── name1   # pool 2, wanted another one, but the name is the same (e.g. misconfiguration)
├── node5
│   └── name1   # pool 2
└── node6
    └── name1   # pool 2
```

Reply:
Get me right. You see, I'm in general picky on name collisions and implicit behavior, because I value human time. And since people tend to copy-and-paste configs from time to time, from machine to machine, they will accidentally make mistakes. And investigating not-that-obvious behavior (from user perspective) is time consuming. So I wish we could report software decisions in logs or anywhere else to avoid possible time loss.

Reply:
In latest commit I added detailed log and additional checks for name collisions, example output:

```
I0322 21:50:03.936197       1 main.go:168] Skip LVM Logical Volume vg0/root as it is not a thin pool
I0322 21:51:20.624090       1 main.go:168] Skip LVM Logical Volume vg0/asdasd as it is not a thin pool
I0322 21:52:04.256049       1 main.go:168] Skip LVM Logical Volume vg0/asdasd as it is not a thin pool
I0322 21:52:37.116028       1 main.go:168] Skip LVM Logical Volume vg0/asdasd as it has no propper tag set: can't find tag with prefix linstor
I0322 21:52:48.036070       1 main.go:168] Skip LVM Logical Volume vg0/asdasd as it has no propper tag set: can't find tag with prefix linstor
I0322 21:53:09.816075       1 main.go:168] Skip LVM Logical Volume vg0/asdasd as it has no propper tag set: found more than one tag with prefix linstor
```